ambry-frontend/src/main/java/com/github/ambry/frontend/PutBlobHandler.java

```diff
@@ -0,0 +1,405 @@
+        restRequest.getMetricsTracker()
+            .injectMetrics(frontendMetrics.postBlobMetricsGroup.getRestRequestMetrics(restRequest.isSslUsed(), true));
+      }
+      byte[] userMetadata = RestUtils.buildUserMetadata(restRequest.getArgs());
+      frontendMetrics.blobPropsBuildTimeInMs.update(System.currentTimeMillis() - propsBuildStartTime);
+      LOGGER.trace("Blob properties of blob being POSTed - {}", blobProperties);
+      return new BlobInfo(blobProperties, userMetadata);
+    }
+
+    /**
+     * Enforce any additional requirements for certain types of uploads like data chunk uploads.
+     * @param blobProperties the {@link BlobProperties} parsed from the request.
+     * @throws RestServiceException
+     */
+    private void checkUploadRequirements(BlobProperties blobProperties) throws RestServiceException {
+      if (RestUtils.isChunkUpload(restRequest.getArgs())) {
+        // ensure that the x-ambry-session header is present.
+        RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.SESSION, true);
+        // validate that a max chunk size is set.
+        RestUtils.getLongHeader(restRequest.getArgs(), RestUtils.Headers.MAX_UPLOAD_SIZE, true);
+        // validate that the TTL for the chunk is set correctly.
+        long chunkTtl = blobProperties.getTimeToLiveInSeconds();
+        if (chunkTtl <= 0 || chunkTtl > frontendConfig.chunkUploadInitialChunkTtlSecs) {
+          throw new RestServiceException("Invalid chunk upload TTL: " + chunkTtl, RestServiceErrorCode.InvalidArgs);
+        }
+      }
+    }
+
+    /**
+     * After {@link SecurityService#processRequest} finishes, call {@link SecurityService#postProcessRequest} to perform
+     * request time security checks that rely on the request being fully parsed and any additional arguments set.
+     * @param blobInfo the {@link BlobInfo} to carry to future stages.
+     * @return a {@link Callback} to be used with {@link SecurityService#processRequest}.
+     */
+    private Callback<Void> securityProcessRequestCallback(BlobInfo blobInfo) {
+      return buildCallback(frontendMetrics.putSecurityProcessRequestMetrics,
+          securityCheckResult -> securityService.postProcessRequest(restRequest,
+              securityPostProcessRequestCallback(blobInfo)), uri, LOGGER, finalCallback);
+    }
+
+    /**
+     * After {@link SecurityService#postProcessRequest} finishes, call {@link Router#putBlob} to persist the blob in the
+     * storage layer.
+     * @param blobInfo the {@link BlobInfo} to make the router call with.
+     * @return a {@link Callback} to be used with {@link SecurityService#postProcessRequest}.
+     */
+    private Callback<Void> securityPostProcessRequestCallback(BlobInfo blobInfo) {
+      return buildCallback(frontendMetrics.postSecurityPostProcessRequestMetrics, securityCheckResult -> {
+        if ("STITCH".equals(RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.PUT_MODE, false))) {
```

Comment:
make a constant for "STITCH", or use an enum

Reply:
Change it to constant.